src/slave/containerizer/mesos/isolators/cgroups2/cgroups2.cpp

```diff
@@ -120,6 +144,569 @@ bool Cgroups2IsolatorProcess::supportsStandalone()
+      continue;
+    }
+
+    Option<ContainerID> containerId = cgroups2_paths::containerId(
+        flags.cgroups_root, cgroup);
+    if (containerId.isNone()) {
+        LOG(INFO) << "Cgroup '" << cgroup << "' does not correspond to a "
+                  << "container id and will not be recovered";
+        continue;
+    }
+
+    if (infos.contains(*containerId)) {
+      // Container has already been recovered.
+      continue;
+    }
+
+    orphans.contains(*containerId) ?
+        knownOrphans.insert(*containerId) :
+        unknownOrphans.insert(*containerId);
+  }
+
+  vector<Future<Nothing>> recovers;
+  foreach (const ContainerID& containerId, knownOrphans) {
+    recovers.push_back(___recover(containerId));
+  }
+
+  foreach (const ContainerID& containerId, unknownOrphans) {
+    recovers.push_back(___recover(containerId));
+  }
+
+  return await(recovers)
+    .then(defer(
+        PID<Cgroups2IsolatorProcess>(this),
+        &Cgroups2IsolatorProcess::__recover,
+        unknownOrphans,
+        lambda::_1));
+}
+
+
+Future<Nothing> Cgroups2IsolatorProcess::__recover(
+    const hashset<ContainerID>& unknownOrphans,
+    const vector<Future<Nothing>>& futures)
+{
+  vector<string> errors;
+  foreach (const Future<Nothing>& future, futures) {
+    if (!future.isReady()) {
+      errors.push_back(future.isFailed() ? future.failure() : "discarded");
+    }
+  }
+  if (errors.size() > 0) {
+    return Failure(
+        "Failed to recover orphan containers: " +
+        strings::join(": ", errors));
+  }
+
+  // Known orphan cgroups will be destroyed by the containerizer using
+  // the normal cleanup path.
+  foreach (const ContainerID& containerId, unknownOrphans) {
+    LOG(INFO) << "Cleaning up unknown orphaned container " << containerId;
+    cleanup(containerId);
+  }
+
+  return Nothing();
+}
+
+
+Future<Nothing> Cgroups2IsolatorProcess::___recover(
+    const ContainerID& containerId)
+{
+  // Check if:
+  // 1. The non-leaf cgroup exists                    => Fail is not found.
+  // 2. The leaf cgroup exists                        => Fail is not found.
+  // 3. Each of the requested controllers are enabled => Log and continue.
+  //
+  // Failure modes:
+  // 1. Container fails during launch.
+  //    This can happen if the launcher fails to `fork`, 'this' isolator fails
+  //    to `prepare` or `isolate`, among other reasons. Cgroups may be
+  //    improperly configured meaning there may be missing cgroups or cgroup
+  //    control files that have the wrong values.
+  // 2. Container fails on destroy.
+  //    The container fails to be destroyed so cgroups may not have been
+  //    correctly cleaned up. This can result in orphan cgroups.
+  // 3. Mesos agent is restarted with different flags.
+  //    If the agent is started with new isolators the cgroups for the existing
+  //    containers, from a previous run, won't have all the requested
+  //    controllers enabled.
```

Comment:
Provide more justification for why it is okay to bail here when there is a missing cgroup or cgroups.

Reply:
Updated to handle invalid container states more gracefully to reduce the likelihood of bailing out.